backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/coverage/CoverageRegionCollector.kt

```diff
@@ -71,32 +71,57 @@ private class IrFunctionRegionsCollector(
         val irFile: IrFile
 ) : IrElementVisitorVoid {
 
+    val regions = mutableMapOf<IrElement, Region>()
+
     private val irFileStack = mutableListOf(irFile)
 
+    private val regionStack = mutableListOf<Region>()
+
+    private val irStatementsStack = mutableListOf<Pair<IrStatement, IrStatement>>()
+
     private val currentFile: IrFile
         get() = irFileStack.last()
 
-    val regions = mutableMapOf<IrElement, Region>()
+    private val currentRegion: Region
+        get() = regionStack.last()
 
     override fun visitElement(element: IrElement) {
         element.acceptChildrenVoid(this)
     }
 
     override fun visitSimpleFunction(declaration: IrSimpleFunction) {
-        declaration.body?.let {
-            recordRegion(it)
-            it.acceptChildrenVoid(this)
+        declaration.body?.let { body ->
+            visitInRegionContext(body) {
+                body.acceptVoid(this)
+            }
         }
     }
 
     override fun visitConstructor(declaration: IrConstructor) {
         val statements = declaration.body?.statements ?: return
-        statements.forEach {
-            if (it is IrDelegatingConstructorCall && !declaration.isPrimary
-                    || it !is IrDelegatingConstructorCall && it !is IrReturn) {
-                recordRegion(it)
-                it.acceptVoid(this)
+        visitInStatementsContext(statements) {
+            if (this is IrDelegatingConstructorCall && !declaration.isPrimary
```

Comment:
Nesting of several `this` is not a good solution from reader's point of view and should be avoided whenever possible. The reason is that it requires reader to remember that this `this` is not pointing to `IrFunctionRegionsCollector` instance but rather to implicit `IrStatement` that came from `visitInStatementsContext`. More obvious code is better :)